File: ai_bot/train_bot.py
import os
import json
import time
import numpy as np
import requests
from dotenv import load_dotenv
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau
from tensorflow.keras.utils import to_categorical
from sklearn.model_selection import train_test_split
import tensorflowjs as tfjs

# 1. 환경 변수 로드 (상위 폴더의 .env 파일 사용)
load_dotenv(dotenv_path='../.env')
SUPABASE_URL = os.environ.get("VITE_SUPABASE_URL")
SUPABASE_KEY = os.environ.get("VITE_SUPABASE_ANON_KEY")

if not SUPABASE_URL or not SUPABASE_KEY:
    raise ValueError("Supabase 환경 변수가 설정되지 않았습니다. .env 파일을 확인해주세요.")

# 설정값
BUCKET_NAME = 'sign-dataset'
FRAMES_PER_SEQUENCE = 30
FEATURES_PER_FRAME = 126 # 양손 (21 * 3 * 2)

def hex_to_string(hex_str):
    """Hex로 인코딩된 폴더명을 원래 한글 단어로 디코딩"""
    try:
        bytes_obj = bytes.fromhex(hex_str)
        return bytes_obj.decode('utf-8')
    except:
        return hex_str

def list_storage(prefix):
    url = f"{SUPABASE_URL}/storage/v1/object/list/{BUCKET_NAME}"
    headers = {"apikey": SUPABASE_KEY, "Authorization": f"Bearer {SUPABASE_KEY}"}
    res = requests.post(url, headers=headers, json={"prefix": prefix, "limit": 1000})
    return res.json() if res.status_code == 200 else []

def download_storage(path):
    url = f"{SUPABASE_URL}/storage/v1/object/public/{BUCKET_NAME}/{path}"
    res = requests.get(url)
    if res.status_code == 200:
        return res.json()
    url_auth = f"{SUPABASE_URL}/storage/v1/object/authenticated/{BUCKET_NAME}/{path}"
    headers = {"apikey": SUPABASE_KEY, "Authorization": f"Bearer {SUPABASE_KEY}"}
    res_auth = requests.get(url_auth, headers=headers)
    if res_auth.status_code == 200:
        return res_auth.json()
    raise Exception(f"Failed to download {path}")

def extract_features(hands_data):
    """프레임 당 손 관절 데이터를 126차원(양손 21관절 x,y,z) 배열로 변환"""
    # 기본값: 0으로 채워진 126차원 배열
    features = np.zeros(FEATURES_PER_FRAME)
    
    if not hands_data or len(hands_data) == 0:
        return features
        
    # 첫 번째 손 (0~62 인덱스)
    hand1 = hands_data[0]
    for i, pt in enumerate(hand1):
        if i >= 21: break
        features[i*3] = pt.get('x', 0)
        features[i*3 + 1] = pt.get('y', 0)
        features[i*3 + 2] = pt.get('z', 0)
        
    # 두 번째 손 (63~125 인덱스)
    if len(hands_data) > 1:
        hand2 = hands_data[1]
        for i, pt in enumerate(hand2):
            if i >= 21: break
            features[63 + i*3] = pt.get('x', 0)
            features[63 + i*3 + 1] = pt.get('y', 0)
            features[63 + i*3 + 2] = pt.get('z', 0)
            
    return features

def download_and_preprocess():
    print("🚀 Supabase에서 데이터를 다운로드하고 전처리를 시작합니다...")
    
    res = list_storage('skeletons')
    folders = [item for item in res if item.get('name') and item['name'] != '.emptyFolder']
    
    X = []
    y = []
    word_labels = []
    
    for idx, folder in enumerate(folders):
        hex_name = folder['name']
        word_name = hex_to_string(hex_name)
        word_labels.append(word_name)
        
        print(f"[{idx+1}/{len(folders)}] '{word_name}' 데이터 수집 중...")
        
        files = list_storage(f'skeletons/{hex_name}')
        json_files = [f for f in files if f.get('name') and f['name'].endswith('.json')]
        
        for file in json_files:
            try:
                file_path = f"skeletons/{hex_name}/{file['name']}"
                data = download_storage(file_path)
                
                if len(data) < FRAMES_PER_SEQUENCE:
                    continue
                
                sequence_features = []
                for i in range(FRAMES_PER_SEQUENCE):
                    frame = data[i]
                    hands = frame.get('hands', [])
                    features = extract_features(hands)
                    sequence_features.append(features)
                    
                X.append(sequence_features)
                y.append(idx)
            except Exception as e:
                print(f"Error processing {file['name']}: {e}")
                
    X = np.array(X)
    y = np.array(y)
    
    print(f"\n✅ 데이터 전처리 완료!")
    print(f"총 샘플 수: {len(X)}")
    print(f"단어 클래스 수: {len(word_labels)} ({', '.join(word_labels)})")
    
    y_categorical = to_categorical(y, num_classes=len(word_labels))
    
    with open('word_labels.json', 'w', encoding='utf-8') as f:
        json.dump(word_labels, f, ensure_ascii=False)
        
    return X, y_categorical, word_labels

def build_model(num_classes):
    print("🧠 LSTM 딥러닝 모델을 생성합니다...")
    model = Sequential([
        LSTM(64, return_sequences=True, activation='relu', input_shape=(FRAMES_PER_SEQUENCE, FEATURES_PER_FRAME)),
        Dropout(0.2),
        LSTM(128, return_sequences=True, activation='relu'),
        Dropout(0.2),
        LSTM(64, return_sequences=False, activation='relu'),
        Dense(64, activation='relu'),
        Dense(num_classes, activation='softmax')
    ])
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    model.summary()
    return model

def run_training():
    start_time = time.time()
    
    X, y, word_labels = download_and_preprocess()
    
    if len(X) == 0:
        print("❌ 학습할 데이터가 없습니다. 훈련을 건너뜁니다.")
        return
        
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    model = build_model(len(word_labels))
    
    early_stopping = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)
    reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=5, min_lr=0.0001)
    
    print("\n⏳ 학습을 시작합니다...")
    model.fit(
        X_train, y_train,
        validation_data=(X_test, y_test),
        epochs=100,
        batch_size=8,
        callbacks=[early_stopping, reduce_lr],
        verbose=1
    )
    
    loss, accuracy = model.evaluate(X_test, y_test)
    print(f"\n📊 최종 평가 결과 - 정확도: {accuracy*100:.2f}%, 손실: {loss:.4f}")
    
    print("\n📦 웹사이트용(TF.js) 모델 변환을 시작합니다...")
    export_dir = '../public/model'
    os.makedirs(export_dir, exist_ok=True)
    
    model.save('temp_model.h5')
    
    print("\n📦 TF.js 포맷으로 변환 중...")
    tfjs.converters.save_keras_model(model, export_dir)
    
    if os.path.exists('temp_model.h5'):
        os.remove('temp_model.h5')
        
    if os.path.exists('word_labels.json'):
        import shutil
        shutil.move('word_labels.json', os.path.join(export_dir, 'word_labels.json'))
        
    end_time = time.time()
    minutes = (end_time - start_time) / 60
    print(f"🎉 훈련 및 배포 완료! (소요 시간: {minutes:.2f}분)")
    print(f"저장 위치: {os.path.abspath(export_dir)}")

def get_dataset_state():
    """클라우드에서 현재 단어별 데이터 세트(JSON) 개수를 파악합니다."""
    res = list_storage('skeletons')
    folders = [item for item in res if item.get('name') and item['name'] != '.emptyFolder']
    
    state = {}
    for folder in folders:
        hex_name = folder['name']
        files = list_storage(f'skeletons/{hex_name}')
        json_count = len([f for f in files if f.get('name') and f['name'].endswith('.json')])
        state[hex_name] = json_count
    return state

def main():
    print("🤖 AI 훈련 봇이 깨어났습니다. (10분마다 새로운 데이터 감지 모드)")
    last_state = {}
    
    while True:
        try:
            print("\n🔍 클라우드 데이터 상태 확인 중...")
            current_state = get_dataset_state()
            should_train = False
            
            for hex_name, count in current_state.items():
                word_name = hex_to_string(hex_name)
                # 데이터가 10세트 이상이고, 이전 기록보다 늘어났다면 훈련 트리거!
                if count >= 10:
                    last_count = last_state.get(hex_name, 0)
                    if count > last_count:
                        print(f"💡 감지됨: '{word_name}' 단어의 데이터가 추가되었습니다! (현재 {count}세트)")
                        should_train = True
                        
            if should_train:
                print("🚀 새로운 데이터가 충분히 모였습니다. 자동 학습을 시작합니다!")
                run_training()
                last_state = current_state
                print("✅ 학습 및 배포 완료! 다시 감시 모드로 들어갑니다.")
            else:
                print("💤 새로운 학습 조건(10세트 이상 변경됨)을 만족하는 데이터가 없습니다.")
                
        except Exception as e:
            print(f"⚠️ 에러 발생 (재시도 대기): {e}")
            
        print("⏳ 10분(600초) 후 다시 검사합니다...")
        time.sleep(600)

if __name__ == "__main__":
    main()

File: ai_bot/train_bot.js
import fs from 'fs';
import path from 'path';
import dotenv from 'dotenv';
import { createClient } from '@supabase/supabase-js';
import * as tf from '@tensorflow/tfjs';

// .env 파일 로드 (루트 폴더)
dotenv.config({ path: path.resolve(process.cwd(), '.env') });

const supabaseUrl = process.env.VITE_SUPABASE_URL;
const supabaseAnonKey = process.env.VITE_SUPABASE_ANON_KEY;

if (!supabaseUrl || !supabaseAnonKey) {
  console.error("❌ 에러: .env 파일에 VITE_SUPABASE_URL 또는 VITE_SUPABASE_ANON_KEY가 설정되어 있지 않습니다.");
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseAnonKey);
const BUCKET_NAME = 'sign-dataset';

// 모델 설정
const FRAMES_PER_SEQUENCE = 30;
const FEATURES_PER_FRAME = 126;

// 저장 경로
const PUBLIC_MODEL_DIR = path.resolve(process.cwd(), 'public/model');

// Hex to String 디코딩
const decodeHex = (hexStr) => {
  try {
    if (/^[0-9a-fA-F]+$/.test(hexStr)) {
      return Buffer.from(hexStr, 'hex').toString('utf8');
    }
  } catch (e) {}
  return hexStr;
};

// 프레임 당 손 관절 데이터를 126차원(양손 21관절 x,y,z) 배열로 변환
const extractFeatures = (hands) => {
  const features = new Array(FEATURES_PER_FRAME).fill(0);
  if (!hands || hands.length === 0) return features;
  
  const hand1 = hands[0];
  for (let i = 0; i < 21; i++) {
    if (hand1[i]) {
      features[i * 3] = hand1[i].x || 0;
      features[i * 3 + 1] = hand1[i].y || 0;
      features[i * 3 + 2] = hand1[i].z || 0;
    }
  }
  
  if (hands.length > 1) {
    const hand2 = hands[1];
    for (let i = 0; i < 21; i++) {
      if (hand2[i]) {
        features[63 + i * 3] = hand2[i].x || 0;
        features[63 + i * 3 + 1] = hand2[i].y || 0;
        features[63 + i * 3 + 2] = hand2[i].z || 0;
      }
    }
  }
  return features;
};

// 현재 클라우드의 데이터 상태를 가져오는 함수 (모니터링용)
async function getDatasetState() {
  const { data: folders, error } = await supabase.storage.from(BUCKET_NAME).list('skeletons', { limit: 1000 });
  if (error) throw error;
  
  const state = {};
  for (const folder of folders) {
    if (!folder.name || folder.name === '.emptyFolder' || folder.metadata) continue;
    const { data: files } = await supabase.storage.from(BUCKET_NAME).list(`skeletons/${folder.name}`, { limit: 1000 });
    const jsonCount = files ? files.filter(f => f.metadata && f.name.endsWith('.json')).length : 0;
    state[folder.name] = jsonCount;
  }
  return state;
}

// 데이터 다운로드 및 전처리 함수 (스마트 캐싱 적용)
const DATASET_CACHE_DIR = path.resolve(process.cwd(), 'ai_bot/dataset_cache');

async function downloadAndPreprocess() {
  console.log("🚀 클라우드 데이터 동기화 및 전처리를 시작합니다 (스마트 캐싱 가동)...");

  if (!fs.existsSync(DATASET_CACHE_DIR)) {
    fs.mkdirSync(DATASET_CACHE_DIR, { recursive: true });
  }

  const { data: folders, error: rootError } = await supabase.storage.from(BUCKET_NAME).list('skeletons', { limit: 1000 });
  if (rootError) throw rootError;

  const validFolders = folders.filter(f => f.name && f.name !== '.emptyFolder' && !f.metadata);
  
  const X_arr = [];
  const y_arr = [];
  const word_labels = [];
  let downloadedCount = 0;
  let cachedCount = 0;

  for (let i = 0; i < validFolders.length; i++) {
    const hexName = validFolders[i].name;
    const wordName = decodeHex(hexName);
    word_labels.push(wordName);
    
    // 워드별 로컬 폴더 확인 및 생성
    const wordCacheDir = path.join(DATASET_CACHE_DIR, hexName);
    if (!fs.existsSync(wordCacheDir)) {
      fs.mkdirSync(wordCacheDir, { recursive: true });
    }
    
    console.log(`[${i + 1}/${validFolders.length}] '${wordName}' 데이터 동기화 중...`);
    
    const { data: files } = await supabase.storage.from(BUCKET_NAME).list(`skeletons/${hexName}`, { limit: 1000 });
    if (!files) continue;

    const jsonFiles = files.filter(f => f.metadata && f.name.endsWith('.json'));
    
    for (const file of jsonFiles) {
      try {
        const localFilePath = path.join(wordCacheDir, file.name);
        let text;
        
        // 캐시 확인 로직
        if (fs.existsSync(localFilePath)) {
          // 로컬 캐시에서 즉시 읽기
          text = fs.readFileSync(localFilePath, 'utf8');
          cachedCount++;
        } else {
          // 캐시에 없으면 클라우드에서 다운로드
          const { data: blob, error: downloadError } = await supabase.storage.from(BUCKET_NAME).download(`skeletons/${hexName}/${file.name}`);
          if (downloadError) throw downloadError;
          
          text = await blob.text();
          fs.writeFileSync(localFilePath, text, 'utf8'); // 로컬에 영구 저장
          downloadedCount++;
        }
        
        const data = JSON.parse(text);
        
        if (data.length < FRAMES_PER_SEQUENCE) continue;
        
        const sequenceFeatures = [];
        for (let j = 0; j < FRAMES_PER_SEQUENCE; j++) {
          const frame = data[j];
          const hands = frame.hands || [];
          sequenceFeatures.push(extractFeatures(hands));
        }
        
        X_arr.push(sequenceFeatures);
        y_arr.push(i); // label index
      } catch (err) {
        console.error(`Error processing ${file.name}: ${err.message}`);
      }
    }
  }

  console.log(`\n✅ 데이터 전처리 완료!`);
  console.log(`📦 이번에 새로 다운받은 파일: ${downloadedCount}개`);
  console.log(`⚡ 로컬 캐시에서 고속으로 읽은 파일: ${cachedCount}개`);
  console.log(`총 샘플 수: ${X_arr.length}`);
  console.log(`단어 클래스 수: ${word_labels.length} (${word_labels.join(', ')})`);

  return { X_arr, y_arr, word_labels };
}

// Node.js용 커스텀 모델 저장 핸들러 (순수 fs 기반)
const getLocalSaveHandler = (dirPath) => {
  return {
    save: async (modelArtifacts) => {
      if (!fs.existsSync(dirPath)) {
        fs.mkdirSync(dirPath, { recursive: true });
      }

      // 1. model.json 저장
      const modelJsonPath = path.join(dirPath, 'model.json');
      const modelJson = {
        format: modelArtifacts.format,
        generatedBy: modelArtifacts.generatedBy || `TensorFlow.js v${tf.version_core}`,
        convertedBy: null,
        modelTopology: modelArtifacts.modelTopology,
        weightsManifest: [
          {
            paths: ['weights.bin'],
            weights: modelArtifacts.weightSpecs
          }
        ]
      };
      fs.writeFileSync(modelJsonPath, JSON.stringify(modelJson, null, 2), 'utf8');

      // 2. weights.bin 저장
      if (modelArtifacts.weightData) {
        const weightsBinPath = path.join(dirPath, 'weights.bin');
        const buffer = Buffer.from(modelArtifacts.weightData);
        fs.writeFileSync(weightsBinPath, buffer);
      }

      return {
        modelArtifactsInfo: {
          dateSaved: new Date(),
          modelTopologyType: 'JSON',
          modelTopologyBytes: JSON.stringify(modelArtifacts.modelTopology).length,
          weightSpecsBytes: JSON.stringify(modelArtifacts.weightSpecs).length,
          weightDataBytes: modelArtifacts.weightData ? modelArtifacts.weightData.byteLength : 0
        }
      };
    }
  };
};

// 메인 학습 함수
async function runTraining() {
  const startTime = Date.now();
  
  const { X_arr, y_arr, word_labels } = await downloadAndPreprocess();
  
  if (X_arr.length === 0) {
    console.log("❌ 학습할 데이터가 없습니다. 훈련을 건너뜁니다.");
    return;
  }
  
  const numClasses = word_labels.length;
  
  // 텐서로 변환
  const X = tf.tensor3d(X_arr, [X_arr.length, FRAMES_PER_SEQUENCE, FEATURES_PER_FRAME]);
  const y = tf.oneHot(tf.tensor1d(y_arr, 'int32'), numClasses);

  console.log("🧠 LSTM 딥러닝 모델을 생성합니다...");
  
  const model = tf.sequential();
  model.add(tf.layers.lstm({
    units: 64,
    returnSequences: true,
    activation: 'relu',
    inputShape: [FRAMES_PER_SEQUENCE, FEATURES_PER_FRAME]
  }));
  model.add(tf.layers.dropout({ rate: 0.2 }));
  model.add(tf.layers.lstm({
    units: 128,
    returnSequences: true,
    activation: 'relu'
  }));
  model.add(tf.layers.dropout({ rate: 0.2 }));
  model.add(tf.layers.lstm({
    units: 64,
    returnSequences: false,
    activation: 'relu'
  }));
  model.add(tf.layers.dense({ units: 64, activation: 'relu' }));
  model.add(tf.layers.dense({ units: numClasses, activation: 'softmax' }));

  model.compile({
    optimizer: 'adam',
    loss: 'categoricalCrossentropy',
    metrics: ['accuracy']
  });

  model.summary();

  console.log("\n⏳ 훈련을 시작합니다...");
  
  await model.fit(X, y, {
    epochs: 100,
    batchSize: 8,
    validationSplit: 0.2,
    callbacks: tf.callbacks.earlyStopping({ monitor: 'val_loss', patience: 10 }),
    verbose: 1
  });

  console.log("\n📦 웹사이트용(TF.js) 모델로 저장을 시작합니다...");
  
  // 로컬 파일시스템에 저장
  await model.save(getLocalSaveHandler(PUBLIC_MODEL_DIR));
  
  // 레이블 정보 저장
  fs.writeFileSync(
    path.join(PUBLIC_MODEL_DIR, 'word_labels.json'), 
    JSON.stringify(word_labels, null, 2), 
    'utf8'
  );

  // 텐서 메모리 해제
  X.dispose();
  y.dispose();

  const minutes = (Date.now() - startTime) / 60000;
  console.log(`🎉 훈련 및 배포 완료! (소요 시간: ${minutes.toFixed(2)}분)`);
  console.log(`저장 위치: ${PUBLIC_MODEL_DIR}`);
}

// 감시 모드 실행 (이벤트 기반 무전기 수신 모드)
let isTraining = false;

function startWatchMode() {
  console.log("🤖 AI 훈련 봇이 깨어났습니다. (Admin의 훈련 가동 신호 대기 중...)");
  
  const channel = supabase.channel('training_channel');
  
  channel.on('broadcast', { event: 'trigger_training' }, async (payload) => {
    console.log(`\n💡 [${new Date().toLocaleTimeString()}] 프론트엔드로부터 훈련 시작 신호를 수신했습니다!`);
    
    if (isTraining) {
      console.log("⚠️ 이미 훈련이 진행 중입니다. 중복 실행을 무시합니다.");
      return;
    }
    
    isTraining = true;
    try {
      await runTraining();
      console.log("✅ 수동 트리거 학습 완료! 다시 대기 모드로 돌아갑니다.");
    } catch (err) {
      console.error(`⚠️ 학습 중 에러 발생: ${err.message}`);
    } finally {
      isTraining = false;
    }
  }).subscribe((status) => {
    if (status === 'SUBSCRIBED') {
      console.log("📡 Supabase 실시간 무전(Broadcast) 채널에 성공적으로 연결되었습니다.");
    }
  });
}

// 메인 실행 로직 분기
const isWatchMode = process.argv.includes('--watch');
if (isWatchMode) {
  startWatchMode();
} else {
  runTraining().catch(console.error);
}
